executor/index_merge_reader.go

```diff
@@ -0,0 +1,724 @@
+		if err != nil {
+			e.feedbacks[workID].Invalidate()
+		}
+		cancel()
+		if err := result.Close(); err != nil {
+			logutil.Logger(ctx).Error("close Select result failed:", zap.Error(err))
+		}
+		e.ctx.StoreQueryFeedback(e.feedbacks[workID])
+		e.partialWorkerWg.Done()
+	}()
+
+	return nil
+}
+
+func (e *IndexMergeReaderExecutor) buildPartialTableReader(ctx context.Context, workID int) Executor {
+	tableReaderExec := &TableReaderExecutor{
+		baseExecutor: newBaseExecutor(e.ctx, e.schema, stringutil.MemoizeStr(func() string { return e.id.String() + "_tableReader" })),
+		table:        e.table,
+		dagPB:        e.dagPBs[workID],
+		startTS:      e.startTS,
+		streaming:    e.partialStreamings[workID],
+		feedback:     statistics.NewQueryFeedback(0, nil, 0, false),
+		plans:        e.partialPlans[workID],
+		ranges:       e.ranges[workID],
+	}
+	return tableReaderExec
+}
+
+func (e *IndexMergeReaderExecutor) startPartialTableWorker(ctx context.Context, fetchCh chan<- *lookupTableTask, workID int) {
+	partialTableReader := e.buildPartialTableReader(ctx, workID)
+	err := partialTableReader.Open(ctx)
+	if err != nil {
+		logutil.Logger(ctx).Error("open Select result failed:", zap.Error(err))
+	}
+	tableInfo := e.partialPlans[workID][0].(*plannercore.PhysicalTableScan).Table
+	worker := &partialTableWorker{
+		keepOrder:    e.keepOrders[workID],
+		batchSize:    e.maxChunkSize,
+		maxBatchSize: e.ctx.GetSessionVars().IndexLookupSize,
+		maxChunkSize: e.maxChunkSize,
+		tableReader:  partialTableReader,
+		tableInfo:    tableInfo,
+	}
+
+	if worker.batchSize > worker.maxBatchSize {
+		worker.batchSize = worker.maxBatchSize
+	}
+	go func() {
+		ctx1, cancel := context.WithCancel(ctx)
+		_, err := worker.fetchHandles(ctx1, fetchCh, e.resultCh, e.finished)
+		if err != nil {
+			e.feedbacks[workID].Invalidate()
+		}
+		cancel()
+		if err := worker.tableReader.Close(); err != nil {
+			logutil.Logger(ctx).Error("close Select result failed:", zap.Error(err))
+		}
+		e.ctx.StoreQueryFeedback(e.feedbacks[workID])
+		e.partialWorkerWg.Done()
+	}()
+
+}
+
+type partialTableWorker struct {
+	keepOrder    bool
+	batchSize    int
+	maxBatchSize int
+	maxChunkSize int
+	tableReader  Executor
+	tableInfo    *model.TableInfo
+}
+
+func (w *partialTableWorker) fetchHandles(ctx context.Context, fetchCh chan<- *lookupTableTask, resultCh chan<- *lookupTableTask, finished <-chan struct{}) (count int64, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			buf := make([]byte, 4096)
+			stackSize := runtime.Stack(buf, false)
+			buf = buf[:stackSize]
+			logutil.Logger(ctx).Error("partialTableWorker in IndexMergeReaderExecutor panicked", zap.String("stack", string(buf)))
+			err4Panic := errors.Errorf("%v", r)
+			doneCh := make(chan error, 1)
+			doneCh <- err4Panic
+			resultCh <- &lookupTableTask{
+				doneCh: doneCh,
+			}
+			if err != nil {
+				err = errors.Trace(err4Panic)
+			}
```

Comment:
How about creating a new function of this logic to handle panic which can reduce some duplicated code.